### ISSUE TITLE
Support of Postgres Copy?

### ISSUE THREAD

**Issue by kanevk** [zdennis/activerecord-import#634]
Postgres has the options to insert a massive volume of data using [Copy](https://www.postgresql.org/docs/9.4/populate.html)

I think using Copy will be a great performance boost for the Gem when used via the PG adapter. Thoughts?

**Comment by jkowens** [comment]
I'm sure Copy is faster, it would be interesting to see how much of a difference it makes. One thing I don't think it supports though is the upsert functionality. For that reason, I think we'd probably have to stick with the current insert statement, but if there are ideas how this might fit in please feel free to share.

**Comment by kanevk** [comment]
@jkowens I could say from personal experience that "Copy" is much much faster for big volumes of data. I will try to provide some benchmarks.

About the interface, you're right, there isn't *not hacky* way to do upsert via `Copy`.

I can see some interface changing options that are reasonable:
 1. Introduce an option `use_copy: [true|false]` with default value `false`.
 This name is Postgres specific and could be used only if the DB is Postgres and if the parameter `on_duplicate_key_update` isn't provided.
 2. Introduce another method in the API called `import_via_copy`(this is pretty explicit) or if we want to be for MySQL as well(LOAD DATA) it should be something more generic like `fast_import` 😄 
 3. Split Insert and Upsert functionalities. Leaving the method `import` only for inserts gives us the possibility to use `Copy`|`LOAD DATA` always if the DB version allows it and use `Insert` otherwise. The upsert functionality may go to another method `upsert` or in more AR-style `bulk_create_or_update`.

(1) is great because is the least interface change and it can show if this functionality is used/liked. There will be a couple of if/elses as checks are the options compatible, is the DB Postgres and is the version supports Copy.
(2) isn't a breaking change as well but the interface is strange in away.

I prefer option (3) the most since it's I like having the separation between insert/upsert functionalities.
I guess there are still under `import`'s hood because there hasn't been big enough difference in their implementations, but now the newest versions of Postgres and MySQL support faster ways for bulk insert and it worth having the separation `.import`/`.upsert`.

**Comment by preetsethi** [comment]
tl;dr ->
1. We initially used `activerecord-import` for our primary data ingestion process, but don't anymore. _(We still use `import` in a bunch of other cases)_
2. Check out https://github.com/diogob/postgres-copy 
3. I can validate that Postgres COPY was significantly more performant in our use case _(importing large CSV files)_.
4. In _most_ production cases _(remote DB, no access to DB filesystem)_, the COPY command is only feasible when streaming data into STDIN.
5. I believe the _OVERALL_ performance gains would be _questionable_ if this feature was to be added, _in most cases that I can imagine_.

--- 

Just wanted to share my long and convoluted experience that's related to this _(aka painful bugs that have paged and woken me up at night)_.  I have experimented with this library and https://github.com/diogob/postgres-copy quite a bit, for a while now. Through all of this experience, I know deep in my soul that we're nowhere near the ideal solution.

### Assume the following (our business objectives)
1. users can define arbitrary data stores
2. users can define fields they care about
3. users can define arbitrary import aliases for these fields
4. users can upload CSV files
5. **users can import data from files, mapping alias fields to defined fields, into a given data store**
6. users can specify fields to _deduplicate_ on and an alternative data store for redundant data

### Background 
1. **CSV files are king.** They're usually **HUGE** in our business domain. **20-100MB** each. Each user aims to process **hundreds** of files a day. _(For a moment, I'm going to ignore how much pain I've had processing large CSVs in Ruby given a memory constrained environment)_.
2. The files have **variable headers** _(if any headers at all)_. 
3. The files have variable encodings and variable line endings _(everything you can imagine: Windows vs *nix line endings, ISO-8859-1 vs UTF-8 encodings)_.

We run on Heroku and we're cheap, always trying to run on the cheapest dynos which are memory constrained (512MB).

### Our current approach (and details about past iterations)
1. We have a pre-process job that uses system-level utilities (cmd line) to read the raw CSV, with various handlers for shitty and unique files (i.e. files i hate). We save files ensuring Postgres copy compatibility _(UTF-8)_. We also save various metadata (number of columns, values for the first row for header determination, etc) for subsequent use.
2. We use `postgres-copy` to insert the pre-processed CSV into a wildly abstract model called `TempCsv` _(essentially what would be a temp table but we tried to stay within the ActiveRecord paradigm to avoid the combination of AR + raw queries + Ruby file IO streaming)_ . Supports up to 30 columns, creatively named `field1` through `field30`.

**We do this because:** In our experience, even using a WIDE spectrum of Ruby CSV libraries and `activerecord-import` with validations _**off**_, `postgres-copy` is always faster. **Significantly**.

_And that makes sense._

The idea of processing these large CSVs in memory (even using chunking) and then using `activerecord-import` is essentially a two step process. For ages, we imported the CSV into Ruby memory space (using a crazy variety of CSV libraries and IO streaming techniques) and then wrote to Postgres using `import`. It worked and was sufficient _for a long time_. 
Any time we had problems: we threw more concurrency at our background worker queue. 
We were always constrained by the CSV processing library and loading raw data into memory. 
I would never fault `activerecord-import`. 
Even when we used IO streaming techniques / chunking, our constraint then became concurrent throughput per Heroku dyno (remember... _we're cheap_). 
We were never constrained by Postgres.

Eventually, our business grew, and so did each user's demand of the system's processing capability. We experimented with `postgres-copy` vs a ton of CSV processing libraries. Obviously `postgres-copy` had more strict requirements regarding format / encoding. It's an all or nothing import: even one poorly encoded character failed the entire copy. But in every case, `postgres-copy` beat out any CSV processing library we found. 

Just to make this clear: *this is not an indictment  of `activerecord-import`*. This library was never our issue.

---

The reason I wrote this long ass post is mainly my _**uncertainty**_. I imagine most people use `activerecord-import` for importing data from files. I could be wildly off base, and my thinking might be limited to my environmental constraints, but would the overall gains using `COPY` really improve performance when you're loading data into Ruby memory space? Using `postgres-copy` allows you to stream raw data directly to Postgres _(given that this raw data is sufficiently formatted / might require preprocessing)_

I would love to hear opposing views and other experience, please share if you disagree. If you save me from getting woken up by PagerDuty, I'll gladly accept being wrong and buy you a beer.

**Comment by jkowens** [comment]
@preetsethi your concluding thought is exactly what I have been thinking. If someone is able to make use of Postgres `COPY`, the [postgres-copy](https://github.com/diogob/postgres-copy) gem is the way to go.
